src/clj/puppetlabs/services/jruby/jruby_puppet_agents.clj

```diff
@@ -65,62 +65,101 @@
   "Flush a single JRuby instance.  Create a new replacement instance
   and insert it into the specified pool."
   [pool-context :- jruby-schemas/PoolContext
-   {:keys [scripting-container jruby-puppet id pool] :as instance} :- JRubyPuppetInstance
+   instance :- JRubyPuppetInstance
    new-pool :- jruby-schemas/pool-queue-type
    new-id   :- schema/Int
    config   :- jruby-schemas/JRubyPuppetConfig
    profiler :- (schema/maybe PuppetProfiler)]
-  (.unregister pool instance)
-  (.terminate jruby-puppet)
-  (.terminate scripting-container)
-  (log/infof "Cleaned up old JRuby instance with id %s, creating replacement."
-             id)
+  (jruby-internal/cleanup-pool-instance! instance)
   (jruby-internal/create-pool-instance! new-pool new-id config
                                         (partial send-flush-instance! pool-context)
                                         profiler))
 
 (schema/defn ^:always-validate
-  flush-pool!
-  "Flush of the current JRuby pool.  NOTE: this function should never
-  be called except by the pool-agent."
-  [pool-context :- jruby-schemas/PoolContext]
-  ;; Since this function is only called by the pool-agent, and since this
-  ;; is the only entry point into the pool flushing code that is exposed by the
-  ;; service API, we know that if we receive multiple flush requests before the
-  ;; first one finishes, they will be queued up and the body of this function
-  ;; will be executed atomically; we don't need to worry about race conditions
-  ;; between the steps we perform here in the body.
-  (log/info "Flush request received; creating new JRuby pool.")
+  pool-initialized? :- schema/Bool
+  "Determine if the current pool has been fully initialized."
+  [expected-pool-size :- schema/Int
+   pool :- jruby-schemas/pool-queue-type]
+  (= expected-pool-size (count (.getRegisteredElements pool))))
+
+(schema/defn ^:always-validate
+  swap-and-drain-pool!
+  "Replace the current pool with a new pool and drain the old pool,
+  optionally refilling the new pool with fresh jrubies."
+  [pool-context :- jruby-schemas/PoolContext
+   old-pool :- jruby-schemas/PoolState
+   new-pool-state :- jruby-schemas/PoolState
+   refill :- schema/Bool]
```

Comment:
And, it'd be nice if we could refactor this out even further and get rid of this parameter (definitely a code smell to have a boolean parameter which toggles significant functionality on/off) but that said, I think this looks a lot better and it'd probably be something of a pain to keep going, so :+1: